### PR TITLE
Compatibility test

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
As  described in this [commit](https://lore.kernel.org/netdev/20210521234227.1286058-1-andrii@kernel.org/T/), when `libbpf 1.0` is  released, `libbpf` will work with `LIBBPF_STRICT_ALL`, we are adding a call for all functional testers that we have to be sure we are already prepared for changes that are coming. 

We are also adding a script that help us to test everything.
##### Test Plan

1. Clone branch
2. Go to `co-re` directory and compile: 
```sh
# cd co-re
# make clean; make
```
3. Enter to `tests` directory  inside `co-re` and run the script that we are adding with this PR.

```sh
# bash run_tests.sh
``` 

4. Test at least one `fiilesystem` running manually:

```sh
# ./filesystem --ext4
``` 

##### Additional information

Run the script [`tools/check-kernel-core.sh`](https://raw.githubusercontent.com/netdata/kernel-collector/master/tools/check-kernel-core.sh)

On kernel `5.16.x` this PR will have one failure, because we need to bring a bug fix for it, this will come next days.

| Distribution | Kernel version | Expected result | Standard Output | Standard Error | File System |Info |
|----------------|----------------------|----------------------|------------------------| ------------------- | --------------- | -------|
|Slackware current | 5.15.15 |  Error to find `sys_sync` | [success_slack.txt](https://github.com/netdata/kernel-collector/files/7919224/success_slack.txt) | [error_slack.txt](https://github.com/netdata/kernel-collector/files/7919228/error_slack.txt) | [slack_fs.txt](https://github.com/netdata/kernel-collector/files/7919230/slack_fs.txt) | Successful  result. |
| Arch Linux 2021.04 | 5.16.1-arch1 | Error to find `sys_sync` and `__set_page_dirty` |  [success_arch.txt](https://github.com/netdata/kernel-collector/files/7919238/success_arch.txt) | [error_arch.txt](https://github.com/netdata/kernel-collector/files/7919240/error_arch.txt) |  [arch_fs.txt](https://github.com/netdata/kernel-collector/files/7919241/arch_fs.txt) | Successful  result. |
| Arch Linux 2021.04 | 5.16.2-arch1 | Error to find `sys_sync` and `__set_page_dirty` | [arch_success.txt](https://github.com/netdata/kernel-collector/files/7928649/arch_success.txt) | [arch_error.txt](https://github.com/netdata/kernel-collector/files/7928651/arch_error.txt) |  [arch_fs.txt](https://github.com/netdata/kernel-collector/files/7928652/arch_fs.txt) |  Successful  result. |
| Ubuntu 21.04 | 5.11.0-49 | Error to find `sys_sync` | [success_ubuntu.txt](https://github.com/netdata/kernel-collector/files/7928659/success_ubuntu.txt) | [error_ubuntu.txt](https://github.com/netdata/kernel-collector/files/7928661/error_ubuntu.txt) |  [ubuntu_fs.txt](https://github.com/netdata/kernel-collector/files/7928662/ubuntu_fs.txt) | Successful result |
| Rocky 8.5 | 4.18.0-348.12.2.el8_5 | It is not expected any error. | [sucess_rocky.txt](https://github.com/netdata/kernel-collector/files/7919256/sucess_rocky.txt) | [error_rocky.txt](https://github.com/netdata/kernel-collector/files/7919260/error_rocky.txt) | [arch_fs.txt](https://github.com/netdata/kernel-collector/files/7919261/arch_fs.txt) | Successful  result. |

### CODE REVIEW

**Comment by erdem2000** [comment]
Hi @thiagoftsm, unfortunately I got those "elf" errors below, do you have any idea?

```
erdem@erdem-VirtualBox:~/netdatadev/kernel-collector/co-re$ make
cd ../libbpf/src && make BUILD_STATIC_ONLY=1 DESTDIR=../../.local_libbpf INCLUDEDIR= LIBDIR= UAPIDIR= install \

make[1]: Entering directory '/home/erdem/netdatadev/kernel-collector/libbpf/src'
  MKDIR    staticobjs
  CC       staticobjs/bpf.o
  CC       staticobjs/btf.o
  CC       staticobjs/libbpf.o
  CC       staticobjs/libbpf_errno.o
  CC       staticobjs/netlink.o
  CC       staticobjs/nlattr.o
  CC       staticobjs/str_error.o
  CC       staticobjs/libbpf_probes.o
  CC       staticobjs/bpf_prog_linfo.o
  CC       staticobjs/xsk.o
  CC       staticobjs/btf_dump.o
  CC       staticobjs/hashmap.o
  CC       staticobjs/ringbuf.o
  CC       staticobjs/strset.o
  CC       staticobjs/linker.o
  CC       staticobjs/gen_loader.o
  CC       staticobjs/relo_core.o
  AR       libbpf.a
  INSTALL  bpf.h libbpf.h btf.h libbpf_common.h libbpf_legacy.h xsk.h bpf_helpers.h bpf_helper_defs.h bpf_tracing.h bpf_endian.h bpf_core_read.h skel_internal.h libbpf_version.h
  INSTALL  ./libbpf.pc
  INSTALL  ./libbpf.a 
make[1]: Leaving directory '/home/erdem/netdatadev/kernel-collector/libbpf/src'
clang  -I../.local_libbpf/ -I../includes/ -I../libbpf/src/ -I./includes/ -I. -I../libbpf/include/uapi/ -ggdb -O2 -target bpf -D__TARGET_ARCH_x86 -idirafter /usr/local/include -idirafter /usr/lib/llvm-10/lib/clang/10.0.0/include -idirafter /usr/include/x86_64-linux-gnu -idirafter /usr/include  -c cachestat.bpf.c -o cachestat.bpf.o
llvm-strip -g cachestat.bpf.o # strip useless DWARF info
bpftool gen skeleton cachestat.bpf.o > includes/cachestat.skel.h
cc -Wall -ggdb -DMY_LINUX_VERSION_CODE=330452 -I../.local_libbpf/ -I../includes/ -I../libbpf/src/ -I./includes/ -I. -I../libbpf/include/uapi/ -c cachestat.c -o cachestat.o
cc -Wall -ggdb -L../.local_libbpf cachestat.o -lelf -lz -lbpf -o tests/cachestat
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `bpf_object__elf_finish':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1316: undefined reference to `elf_end'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_hdr':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3194: undefined reference to `elf64_getshdr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3196: undefined reference to `elf_errmsg'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3196: undefined reference to `elf_ndxscn'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_str':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3145: undefined reference to `elf_strptr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3147: undefined reference to `elf_errmsg'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_name':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3218: undefined reference to `elf_errmsg'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3218: undefined reference to `elf_ndxscn'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_data':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3233: undefined reference to `elf_getdata'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3235: undefined reference to `elf_errmsg'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3235: undefined reference to `elf_ndxscn'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_by_idx':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3159: undefined reference to `elf_getscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3161: undefined reference to `elf_errmsg'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `bpf_object__read_kconfig_file':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1935: undefined reference to `gzopen64'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1944: undefined reference to `gzgets'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1954: undefined reference to `gzclose'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1937: undefined reference to `gzopen64'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sym_str':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3131: undefined reference to `elf_strptr'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_str':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3147: undefined reference to `elf_errmsg'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `elf_sec_by_name':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3174: undefined reference to `elf_nextscn'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `__bpf_object__open':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:6973: undefined reference to `elf_version'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `bpf_object__elf_init':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1357: undefined reference to `elf_begin'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1368: undefined reference to `elf_kind'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1345: undefined reference to `elf_memory'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1374: undefined reference to `gelf_getclass'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1380: undefined reference to `elf64_getehdr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1387: undefined reference to `elf_getshdrstrndx'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1395: undefined reference to `elf_getscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1395: undefined reference to `elf_rawdata'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1382: undefined reference to `elf_errmsg'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `bpf_object__elf_collect':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3337: undefined reference to `elf_nextscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3352: undefined reference to `elf_ndxscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3367: undefined reference to `elf_nextscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:3368: undefined reference to `elf_ndxscn'
/usr/bin/ld: ../.local_libbpf/libbpf.a(libbpf.o): in function `bpf_object__elf_init':
/home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1361: undefined reference to `elf_errmsg'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1388: undefined reference to `elf_errmsg'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/libbpf.c:1396: undefined reference to `elf_errmsg'
/usr/bin/ld: ../.local_libbpf/libbpf.a(btf.o): in function `btf_parse_elf':
/home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:899: undefined reference to `elf_version'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:913: undefined reference to `elf_begin'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:918: undefined reference to `gelf_getehdr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:923: undefined reference to `elf_getshdrstrndx'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:929: undefined reference to `elf_getscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:929: undefined reference to `elf_rawdata'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:934: undefined reference to `elf_nextscn'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:939: undefined reference to `gelf_getshdr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:944: undefined reference to `elf_strptr'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:959: undefined reference to `elf_getdata'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:951: undefined reference to `elf_getdata'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:980: undefined reference to `gelf_getclass'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
/usr/bin/ld: /home/erdem/netdatadev/kernel-collector/libbpf/src/btf.c:1002: undefined reference to `elf_end'
collect2: error: ld returned 1 exit status
make: *** [Makefile:55: cachestat] Error 1
rm cachestat.bpf.o
erdem@erdem-VirtualBox:~/netdatadev/kernel-collector/co-re$ 
```

**Comment by vlvkobal** [comment]
@erdem2000 is `libelf` installed on your system? Is the `libbpf` submodule updated?

**Review by vlvkobal — APPROVED** [review]

**Comment by erdem2000** [comment]
> @erdem2000 is `libelf` installed on your system? Is the `libbpf` submodule updated?

Hi @vlvkobal, yes, libelf is installed and libbpf info is below;
```
erdem@erdem-VirtualBox:~/netdatadev/kernel-collector/libbpf$ git log -1
commit 4cea98d24104334f2fee1abf23564e1910630f3e (HEAD, tag: v0.6.1_netdata, origin/libbpf_0_6_1)
Author: Thiago Marques <thiagoftsm@gmail.com>
Date:   Sun Dec 5 20:05:09 2021 +0000

    libbpf_0_6_1: Apply patch to keep compatiblity with legacy kernels
```

**Comment by thiagoftsm** [comment]
> @erdem2000 is `libelf` installed on your system? Is the `libbpf` submodule updated?

A complement for previous questions, what is your `libelf` version?

**Comment by thiagoftsm** [comment]
@vlvkobal and @erdem2000 I discovered what was happening on Ubuntu `21.04`. On this distribution we have `clang 12.0`, and we also had a problem to identify a kernel version. So I pushed two commits to fix the issues.

I also udpated the OP with my results on Ubuntu and the latest kernel that Arch is running.

**Review by erdem2000 — APPROVED** [review]
After last commit (2f336541acb27a74b931e1338c7d9de1629b369f) PR is worked as expected.
OS: Ubuntu 20.04
libelf version: 0.176-1.1build1

File: co-re/Makefile [hunk 1]
```diff
@@ -11,8 +11,8 @@ KERNEL_VERSION="$(shell cat /proc/sys/kernel/osrelease)"
 FIRST_KERNEL_VERSION=$(shell sh ../tools/complement.sh "$(KERNEL_VERSION)")
 VER_MAJOR=$(shell echo $(KERNEL_VERSION) | cut -d. -f1)
 VER_MINOR=$(shell echo $(KERNEL_VERSION) | cut -d. -f2)
-VER_PATCH=$(shell echo $(KERNEL_VERSION) | cut -d. -f3)
-RUNNING_VERSION_CODE=$(shell echo $$(( $(VER_MAJOR) * 65536 + $(VER_MINOR) * 256 + $(VER_PATCH) )) )
+VER_PATCH=$(shell echo $(KERNEL_VERSION) | cut -d. -f3 | cut -d\- -f1)
+RUNNING_VERSION_CODE=$(shell echo $$(( $(VER_MAJOR) * 65536 + $(VER_MINOR) * 256 + $(VER_MINOR))) )
```
Comment (vlvkobal) [inline co-re/Makefile:15]: I believe you changed `VER_PATCH` to `VER_MINOR` by mistake.

Comment (thiagoftsm) [reply co-re/Makefile]: You're right, I fixed with last commit. Thanks!

**Review by vlvkobal — APPROVED** [review]

**Review by iigorkarpov — APPROVED** [review]

**Pull request merged**